python/cugraph/structure/number_map.py

```diff
@@ -0,0 +1,263 @@
+            pass
+
+        def to_vertex_id(df, col_names):
+            # TODO: Fill in implementation
+            pass
+
+        def from_vertex_id(series):
+            # TODO: Fill in implementation
+            pass
+
+    class MultiGPU:
+        def __init__(self, ddf, col_names):
+            # TODO: Fill in implementation
+            pass
+
+        def append(self, ddf, col_names):
+            # TODO: Fill in implementation
+            pass
+
+        def compute(self):
+            # TODO: Fill in implementation
+            pass
+
+        def to_vertex_id():
+            # TODO: Fill in implementation
+            pass
+
+        def from_vertex_id():
+            # TODO: Fill in implementation
+            pass
+
+            
+    def __init__(self):
+        self.implementation = None
+
+    def compute_vals(column_names):
+        """
+        Helper function to compute internal column names based on external column names
+        """
+        return [str(i) for i in range(len(column_names))]
+
+    def from_dataframe(self, df, col_names):
+        """
+        Populate the numbering map with vertices from the specified
+        columns of the provided data frame.
+
+        Parameters
+        ----------
+        df : cudf.DataFrame or dask_cudf.DataFrame
+            Contains a list of external vertex identifiers that will be
+            numbered by the NumberMap class.
+        col_names: list of strings
+            This list of 1 or more strings contain the names
+            of the columns that uniquely identify an external
+            vertex identifier
+        """
+        if self.implementation is not None:
+            raise Exception('NumberMap is already populated')
+
+        if type(df) is cudf.DataFrame:
+            self.implementation = NumberMap.SingleGPU(df, col_names)
+        elif type(df) is dask_cudf.DataFrame:
+            self.implementation = NumberMap.MultiGPU(df, col_names)
+        else:
+            raise Exception('df must be cudf.DataFrame or dask_cudf.DataFrame')
+
+        self.implementation.compute()
+
+    def from_cudf_dataframe(self, df, src_col_names, dst_col_names):
+        """
+        Populate the numbering map with vertices from the specified
+        columns of the provided data frame.
+
+        Parameters
+        ----------
+        df : cudf.DataFrame or dask_cudf.DataFrame
+            Contains a list of external vertex identifiers that will be
+            numbered by the NumberMap class.
+        src_col_names: list of strings
+            This list of 1 or more strings contain the names
+            of the columns that uniquely identify an external
+            vertex identifier for source vertices
+        dst_col_names: list of strings
+            This list of 1 or more strings contain the names
+            of the columns that uniquely identify an external
+            vertex identifier for destination vertices
+        """
```

Comment:
Need to document multi-GPU return behavior 
- Are internal IDs randomly spread or are they grouped together so that all edges for the same source (or perhaps destination) are on the same GPU?
- How does this connect to the next step in the OPG pipeline (coo2csr #812 )? Are they natively compatible or do we need an intermediate step?

Reply:
Regarding the ids, the approach I've been using is as follows:
 * External vertex ids are hashed and that hash value is used to partition vertex ids across the cluster.  Each node of the cluster then is responsible for all vertices that hash to it.
 * Each local node of the cluster now has all of the vertex information, perform a local renumbering to get 0..(n-1) numbering of all vertices that hashed to that node
 * "Somehow" assign global ids.  Current prototype does assignment by computing the number of vertices on each partition and doing a prefix sum to create a base vertex identifier for each node.  The consequence is that we end up with an overall 0..(N-1) mapping, each node getting a contiguous subrange of ids that it created.

Regarding the OPG pipeline, this NumberMap does not entirely implement renumbering, it implements the core capability and will be hidden inside the graph object (although externally callable for those who wish to access the renumbering independently).  The renumbering implementation itself will end up using this to create a COO that should be directly usable as input to #812.

Reply:
Sounds good, I think adding doc about this would be helpful. 
Perhaps with a note as well on getting these properties for CSC as the OPG Pagerank pipeline is the first one?

SG does (1) renumber (2) swap src and dst (3) coo2csr. (1) being at edge list loading time and (2,3) at analytics time. 
Now looking at MG, are we relying on the assumption that (2) gets done before (1) in order to get coo2csr successfully building the local CSC matrices?

Reply:
Can't see the doc for the multi-GPU return behavior, let me know if you added it elsewhere

Reply:
Not sure what you're asking for.  The typical workflow is outlined in the renumber notebooks.  It works like this:

* Instantiate a NumberMap object
* Call from_dataframe to populate the number map
* Call to_internal_vertex_id, add_internal_vertex_id, from_internal_vertex_id as desired to convert between internal and external vertex ids.

Alternatively, you can call NumberMap.renumber on a DataFrame and it will return a fully populated NumberMap and a renumbered DataFrame.  Then you can call NumberMap.unrenumber on a DataFrame and it will convert internal vertex ids back into external vertex ids.

If you call it with a cudf.DataFrame, it uses Single GPU logic to create and translate everything.  If you call it with a dask_cudf.DataFrame it uses MG logic and generated dask_cudf.DataFrame objects everywhere.  The NumberMap object keeps track internally of everything necessary to do this.

Reply:
I further wrapped this in the Graph renumber and unrenumber functions which are automatically used if you specify renumber=True on the from_cudf_edgelist or from_dask_cudf_edgelist calls.

Reply:
> Not sure what you're asking for.

Documentation that describes the data distribution behavior of MG renumbering output in general. Users of the renumbering feature (and MG developer) will wonder if each node gets a contiguous range or is it scattered for instance.